datastore/src/main/java/com/ds/datastore/BookStoreController.java

```diff
@@ -3,78 +3,229 @@
-    protected ResponseEntity<EntityModel<Book>> newBook(@RequestBody Book book, @PathVariable Long store_id){
-        BookStore bookStore = repository.findById(store_id).orElseThrow();
-        BookModelAssembler bookModelAssembler = new BookModelAssembler();
-        EntityModel<Book> entityModel = bookModelAssembler.toModel(book);
-        bookStore.addBook(book);
+    @PostMapping("/bookstores")
+    protected ResponseEntity<EntityModel<BookStore>> newBookStore(@RequestBody BookStore bookStore) throws Exception {
+        if (this.id != null) {
+            return ResponseEntity.status(HttpStatus.ALREADY_REPORTED).build();
+        }
+        EntityModel<BookStore> entityModel = postToHub(bookStore);
+        System.out.println(map.getMap().toString());
         return ResponseEntity
                 .created(entityModel.getRequiredLink(IanaLinkRelations.SELF).toUri())
                 .body(entityModel);
     }
-    */
 
-    @GetMapping("/bookstores/{storeID}")
-    protected EntityModel<BookStore> one(@PathVariable Long storeID) {
-        BookStore bookStore = repository.findById(storeID).orElseThrow(() -> new BookStoreNotFoundException(storeID));
-        return assembler.toModel(bookStore);
+    @PostMapping("/bookstores/{id}")
+    protected void addServer(@RequestBody String json, @PathVariable String id) {
+        JsonObject jso = new JsonParser().parse(json).getAsJsonObject();
+        Long givenID = jso.get("id").getAsLong();
+        String address = jso.get("address").getAsString();
+        this.map.put(givenID, address);
     }
 
+    @GetMapping("/bookstores/{storeID}")
+    protected ResponseEntity one(@PathVariable Long storeID) throws Exception {
+        try{
+            BookStore bookStore = storeRepository.findByServerId(storeID).orElseThrow(() -> new BookStoreNotFoundException(storeID));
+            EntityModel<BookStore> entityModel = assembler.toModel(bookStore);
+            return ResponseEntity
+                    .created(entityModel.getRequiredLink(IanaLinkRelations.SELF).toUri())
+                    .body(entityModel);
+        }catch (BookStoreNotFoundException e){
+            if(this.map.containsKey(storeID)){
+                UriComponentsBuilder builder = UriComponentsBuilder.fromHttpUrl(this.map.get(storeID));
+                URI uri = new URI(builder.toUriString());
+                return ResponseEntity.status(HttpStatus.PERMANENT_REDIRECT).location(uri).build();
+            }else{
+                throw e;
+            }
+        }
+    }
 
     @GetMapping("/bookstores")
-    protected CollectionModel<EntityModel<BookStore>> all(){
-        List<EntityModel<BookStore>> bookStores = repository.findAll().stream()
-                .map(assembler::toModel)
-                .collect(Collectors.toList());
-        return CollectionModel.of(bookStores, linkTo(methodOn(BookStoreController.class).all()).withSelfRel());
+    protected CollectionModel<EntityModel<BookStore>> all() throws Exception {
```

Comment:
looks like you did implement the get all bookstores?

Reply:
Started but we stopped when you told us to quit it. So it isn't functional

Reply:
👍